### PR TITLE
Closes #98

### LINKED ISSUE #98: Aloha H2O model doesn't support options
We need to support the ability to have optional types in the CSV producers.  This translates into H2O models that support options.

### PR DESCRIPTION
**Pull request by deaktator**
This change is [Reviewable](https://reviewable.io/reviews/eharmony/aloha/101)

### CODE REVIEW

File: aloha-core/src/main/scala/com/eharmony/aloha/dataset/csv/CsvRowCreator.scala [hunk 1]
```diff
@@ -92,7 +91,10 @@ final object CsvRowCreator {
                                 case EncodingBasedFinalizer(fnl) => fnl(encoding)
                             }
 
-                            val strFunc = success.andThenGenAggFunc(finalizer)
+                            val strFunc = success.
+                              andThenGenAggFunc(_ orElse csvCol.defVal).  // Issue #98: Inject defVal on None.
```
Comment (deaktator) [inline aloha-core/src/main/scala/com/eharmony/aloha/dataset/csv/CsvRowCreator.scala:95]: This line is the one that allows the default to be returned when the function returns `None`.

File: aloha-core/src/main/scala/com/eharmony/aloha/dataset/csv/json/csvJson.scala [hunk 2]
```diff
@@ -1,18 +1,31 @@
...
+import com.eharmony.aloha.factory.Formats
 import com.eharmony.aloha.reflect.{RefInfo, RefInfoOps}
 import com.eharmony.aloha.util.Logging
 import spray.json._
 
 import scala.collection.{immutable => sci}
 import scala.language.implicitConversions
+import scala.util.Try
 
+/**
```
Comment (deaktator) [inline aloha-core/src/main/scala/com/eharmony/aloha/dataset/csv/json/csvJson.scala:14]: Didn't finish this comment.  Need to do so.

File: aloha-core/src/main/scala/com/eharmony/aloha/dataset/csv/json/csvJson.scala [hunk 3]
```diff
@@ -49,27 +64,65 @@ extends DefaultJsonProtocol
...
                 case Seq(JsString(s)) => Option(s.toLowerCase)
                 case _ => None
             }
 
+            val optional = o.getFields("optional") match {
+                case Seq(JsBoolean(bool)) => bool
+                case _ => false
+            }
+
+            lazy val enumClassName = o.getFields("enumClass") match {
```
Comment (deaktator) [inline aloha-core/src/main/scala/com/eharmony/aloha/dataset/csv/json/csvJson.scala:89]: Added these to make it easier to deal with booleans.

File: aloha-core/src/main/scala/com/eharmony/aloha/dataset/csv/json/csvJson.scala [hunk 4]
```diff
@@ -49,27 +64,65 @@ extends DefaultJsonProtocol
...
+                case Some("enum") =>
+                    enumClassName.map {
+                        case name if optional =>
+                            val clazz = enumClass getOrElse {
+                                throw new DeserializationException(s"class '$name' is not an Enum class.")
+                            }
+                            o.convertTo(OptionEnumCsvColumn.oeccFormat(clazz))
+                        case _ => o.convertTo[EnumCsvColumn]
+                    } orElse {
+                        Option(o.convertTo[SyntheticEnumCsvColumn])
```
Comment (deaktator) [inline aloha-core/src/main/scala/com/eharmony/aloha/dataset/csv/json/csvJson.scala]: Crap.  This isn't done.  Need to add the optional case.

File: aloha-core/src/test/scala/com/eharmony/aloha/semantics/compiled/CompiledSemanticsTest.scala [hunk 5]
```diff
@@ -16,6 +16,25 @@ import org.junit.runners.BlockJUnit4ClassRunner
 class CompiledSemanticsTest {
     private[this] val compiler = TwitterEvalCompiler(classCacheDir = Option(FileLocations.testGeneratedClasses))
 
+//    @Test def testJonMorraTypeAscriptionProblem() {
```
Comment (deaktator) [inline aloha-core/src/test/scala/com/eharmony/aloha/semantics/compiled/CompiledSemanticsTest.scala]: Should remove this eventually.

**Comment by jmorra** [comment]
I think this looks really good.  I don't have anything else to add.

<details><summary>

<i>Previously, deaktator (Ryan Deak) wrote…</i></summary>



> #### Closes #98
> 
> </details>

---

Review status: 0 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/eharmony/aloha/101#-:-KJBcMSbn93N7UFHonrQ:-1873470088)_

**Comment by deaktator** [comment]
Reviewed 6 of 8 files at r1, 3 of 4 files at r2, 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/eharmony/aloha/101)_

**Comment by deaktator** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

_[aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/json/H2oModelJson.scala, line 43 [r4]](https://reviewable.io:443/reviews/eharmony/aloha/101#-KJDCvO4Qq-dFvf1XcSv:-KJDCvO5bMBNqEpGyct-:-53503929) ([raw file](https://github.com/eharmony/aloha/blob/03dc5b2686ab72cfc922ee2725521749f3aa8906/aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/json/H2oModelJson.scala#L43)):_

> ``` Scala
>       val jso = json.asJsObject
>       jso.fields.get("type") match {
>         case None                     => jso.convertTo(jsonFormat3(DoubleH2oSpec)) // Default is double type.
> ```

Not 100% sure that if no `type` is specified but `defVal` is provided that it parses correctly.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/eharmony/aloha/101)_

**Comment by deaktator** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

_[aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/json/H2oModelJson.scala, line 62 [r4]](https://reviewable.io:443/reviews/eharmony/aloha/101#-KJDHkeXzV8mhf8tVkqy:-KJDHkeXzV8mhf8tVkqz:524958891) ([raw file](https://github.com/eharmony/aloha/blob/03dc5b2686ab72cfc922ee2725521749f3aa8906/aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/json/H2oModelJson.scala#L62)):_

> ``` Scala
>           case Some(JsString(d))        => throw new DeserializationException(s"unsupported H2oSpec type: $d. Should be 'double' or 'string'.")
>           case Some(d)                  => throw new DeserializationException(s"H2oSpec type expected string, got: $d")
>           case _                        => throw new DeserializationException(s"No 'type' field present.")
> ```

@jmorra Do we want to allow a type of `"double"`  when it type is omitted? 

``` json
{
  "myCoolFeature": { "spec": "[ SPEC HERE ]", "defVal": 123.456 }, 
```

If so, we need to change this:

``` scala
case _ => throw new DeserializationException(s"No 'type' field present.")
```

to this:

``` scala
case None => (k, DoubleH2oSpec(k, spec(o), o.fields.get("defVal").flatMap(_.convertTo[Option[Double]])))
```

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/eharmony/aloha/101)_

**Comment by jmorra** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_[aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/json/H2oModelJson.scala, line 62 [r4]](https://reviewable.io:443/reviews/eharmony/aloha/101#-KJDHkeXzV8mhf8tVkqy:-KJDWy5TTrR9KOSFry2q:701179957) ([raw file](https://github.com/eharmony/aloha/blob/03dc5b2686ab72cfc922ee2725521749f3aa8906/aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/json/H2oModelJson.scala#L62)):_
<details><summary>

<i>Previously, deaktator (Ryan Deak) wrote…</i></summary>



> @jmorra Do we want to allow a type of `"double"`  when it type is omitted? 
> 
> ``` json
> {
>   "myCoolFeature": { "spec": "[ SPEC HERE ]", "defVal": 123.456 }, 
> ```
> 
> If so, we need to change this:
> 
> ``` scala
> case _ => throw new DeserializationException(s"No 'type' field present.")
> ```
> 
> to this:
> 
> ``` scala
> case None => (k, DoubleH2oSpec(k, spec(o), o.fields.get("defVal").flatMap(_.convertTo[Option[Double]])))
> ```
> 
> </details>
> I think the default should be double.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/eharmony/aloha/101)_

**Comment by deaktator** [comment]
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_[aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/json/H2oModelJson.scala, line 62 [r4]](https://reviewable.io:443/reviews/eharmony/aloha/101#-KJDHkeXzV8mhf8tVkqy:-KJDb9X56wRUqypCDYvp:-499146009) ([raw file](https://github.com/eharmony/aloha/blob/03dc5b2686ab72cfc922ee2725521749f3aa8906/aloha-h2o/src/main/scala/com/eharmony/aloha/models/h2o/json/H2oModelJson.scala#L62)):_
<details><summary>

<i>Previously, jmorra (Jon Morra) wrote…</i></summary>



> I think the default should be double.
> </details>
> Done.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/eharmony/aloha/101)_

**Pull request merged**